clients/include/testing_hpr.hpp

```diff
@@ -0,0 +1,190 @@
+    device_vector<U> d_alpha(1);
+    CHECK_HIP_ERROR(dA_1.memcheck());
+    CHECK_HIP_ERROR(dA_2.memcheck());
+    CHECK_HIP_ERROR(dx.memcheck());
+    CHECK_HIP_ERROR(d_alpha.memcheck());
+
+    double gpu_time_used, cpu_time_used;
+    double rocblas_gflops, cblas_gflops, rocblas_bandwidth;
+    double rocblas_error_1;
+    double rocblas_error_2;
+
+    // Initial Data on CPU
+    rocblas_init(hA_1, true);
+    rocblas_init(hx, false);
+
+    // copy matrix is easy in STL; hA_gold = hA_1: save a copy in hA_gold which will be output of
+    // CPU BLAS
+    hA_gold = hA_1;
+    hA_2    = hA_1;
+
+    // copy data from CPU to device
+    CHECK_HIP_ERROR(dA_1.transfer_from(hA_1));
+    CHECK_HIP_ERROR(dA_2.transfer_from(hA_1));
+    CHECK_HIP_ERROR(dx.transfer_from(hx));
+    CHECK_HIP_ERROR(d_alpha.transfer_from(halpha));
+
+    if(arg.unit_check || arg.norm_check)
+    {
+        CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_host));
+        CHECK_ROCBLAS_ERROR((rocblas_hpr<T, U>)(handle, uplo, N, &h_alpha, dx, incx, dA_1));
+
+        CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_device));
+        CHECK_ROCBLAS_ERROR((rocblas_hpr<T, U>)(handle, uplo, N, d_alpha, dx, incx, dA_2));
+
+        // copy output from device to CPU
+        CHECK_HIP_ERROR(hA_1.transfer_from(dA_1));
+        CHECK_HIP_ERROR(hA_2.transfer_from(dA_2));
+
+        // CPU BLAS
+        cpu_time_used = get_time_us();
+        cblas_hpr<T, U>(uplo, N, h_alpha, hx, incx, hA_gold);
+        cpu_time_used = get_time_us() - cpu_time_used;
+        cblas_gflops  = hpr_gflop_count<T>(N) / cpu_time_used * 1e6;
+
+        if(arg.unit_check)
+        {
+            const double tol = N * sum_error_tolerance<T>;
+            near_check_general<T>(1, size_A, 1, hA_gold, hA_1, tol);
+            near_check_general<T>(1, size_A, 1, hA_gold, hA_2, tol);
+        }
+
+        if(arg.norm_check)
+        {
+            rocblas_error_1 = norm_check_general<T>('F', 1, size_A, 1, hA_gold, hA_1);
+            rocblas_error_2 = norm_check_general<T>('F', 1, size_A, 1, hA_gold, hA_2);
+        }
+    }
+
+    if(arg.timing)
+    {
+        int number_cold_calls = 2;
+        int number_hot_calls  = arg.iters;
+        CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_host));
+
+        for(int iter = 0; iter < number_cold_calls; iter++)
+        {
+            rocblas_hpr<T, U>(handle, uplo, N, &h_alpha, dx, incx, dA_1);
+        }
+
+        gpu_time_used = get_time_us(); // in microseconds
+
+        for(int iter = 0; iter < number_hot_calls; iter++)
+        {
+            rocblas_hpr<T, U>(handle, uplo, N, &h_alpha, dx, incx, dA_1);
+        }
+
+        gpu_time_used     = (get_time_us() - gpu_time_used) / number_hot_calls;
+        rocblas_gflops    = hpr_gflop_count<T>(N) / gpu_time_used * 1e6;
+        rocblas_bandwidth = (2.0 * N * (N + 1)) / 2 * sizeof(T) / gpu_time_used / 1e3;
```

Comment:
shouldn't we have a xxx_bandwith function?

Reply:
I think Torre is working on this.

Reply:
Sorry you need to wait for PR #956 then add to bytes.hpp